src/jdk.jfr/share/classes/jdk/jfr/internal/dcmd/AbstractDCmd.java

```diff
@@ -269,4 +271,41 @@ protected final String exampleDirectory() {
             return "/directory/recordings";
         }
     }
+
+    static String expandFilename(String filename) {
+        if (filename == null || filename.indexOf('%') == -1) {
+            return filename;
+        }
+
+        String pid = null;
+        String time = null;
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < filename.length(); i++) {
+            char c = filename.charAt(i);
+            if (c == '%' && i < filename.length() - 1) {
+                char nc = filename.charAt(i + 1);
+                if (nc == '%') { // %% ==> %
+                    sb.append('%');
+                    i++;
+                } else if (nc == 'p') {
+                    if (pid == null) {
+                        pid = JVM.getJVM().getPid();
+                    }
+                    sb.append(pid);
+                    i++;
+                } else if (nc == 't') {
+                    if (time == null) {
+                        time = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss").format(new Date());
```

Comment:
I use the current time here so that different filenames could be generated for multiple dumps of a process.

Reply:
Good suggestion.
Fixed.